File: backend/app/security.py
import secrets
import hashlib
from datetime import datetime, timedelta, timezone

import bcrypt
import jwt

from app.config import (
    JWT_SECRET,
    JWT_ALGORITHM,
    ACCESS_TOKEN_LIFETIME_MINUTES,
    REFRESH_TOKEN_LIFETIME_DAYS,
)


def verify_password(plain_password: str, password_hash: str) -> bool:
    return bcrypt.checkpw(plain_password.encode(), password_hash.encode())


def hash_password(plain_password: str) -> str:
    return bcrypt.hashpw(plain_password.encode(), bcrypt.gensalt()).decode()


def create_access_token(user_id: str, active_tenant_id: str | None) -> str:
    """
    Note what's deliberately NOT in here: no accessible_tenant_ids[] array.
    Under Option B, authorization is re-checked against the live
    tenant_memberships table on every request — the token only needs to
    say who you are and which tenant you're currently "standing inside."
    """
    now = datetime.now(timezone.utc)
    payload = {
        "user_id": str(user_id),
        "active_tenant_id": str(active_tenant_id) if active_tenant_id else None,
        "iat": now,
        "exp": now + timedelta(minutes=ACCESS_TOKEN_LIFETIME_MINUTES),
    }
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def decode_access_token(token: str) -> dict:
    return jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])


def generate_refresh_token() -> tuple[str, str, datetime]:
    """Returns (raw_token_to_send_to_client, hash_to_store_in_db, expires_at)."""
    raw_token = secrets.token_urlsafe(48)
    token_hash = hashlib.sha256(raw_token.encode()).hexdigest()
    expires_at = datetime.now(timezone.utc) + timedelta(days=REFRESH_TOKEN_LIFETIME_DAYS)
    return raw_token, token_hash, expires_at


def hash_refresh_token(raw_token: str) -> str:
    return hashlib.sha256(raw_token.encode()).hexdigest()

File: apps/backend-node/src/auth/security.ts
import { Injectable } from '@nestjs/common';
import * as bcrypt from 'bcryptjs';
import * as crypto from 'crypto';
import * as jwt from 'jsonwebtoken';
import { ConfigService } from '../config/config.service';

/**
 * Password hashing, HS256 JWT issue/verify, and refresh-token generation.
 * Faithful port of backend/app/security.py.
 *
 * bcrypt hashes ($2a$/$2b$) are portable between the Python `bcrypt` lib and
 * `bcryptjs`, so existing password hashes verify unchanged. Refresh tokens are
 * stored only as SHA-256 hashes; the raw token goes to the client cookie.
 */

export interface AccessTokenClaims {
  user_id: string;
  active_tenant_id: string | null;
  iat?: number;
  exp?: number;
}

export class TokenExpiredError extends Error {}
export class InvalidTokenError extends Error {}

@Injectable()
export class SecurityService {
  constructor(private readonly config: ConfigService) {}

  verifyPassword(plain: string, passwordHash: string): boolean {
    return bcrypt.compareSync(plain, passwordHash);
  }

  hashPassword(plain: string): string {
    return bcrypt.hashSync(plain, bcrypt.genSaltSync());
  }

  /**
   * Note what's deliberately NOT in here: no accessible_tenant_ids[] array.
   * Under Option B, authorization is re-checked against the live DB on every
   * request — the token only says who you are and which tenant you're
   * currently "standing inside."
   */
  createAccessToken(userId: string, activeTenantId: string | null): string {
    const payload: AccessTokenClaims = {
      user_id: String(userId),
      active_tenant_id: activeTenantId ? String(activeTenantId) : null,
    };
    return jwt.sign(payload, this.config.JWT_SECRET, {
      algorithm: this.config.JWT_ALGORITHM,
      expiresIn: this.config.ACCESS_TOKEN_LIFETIME_MINUTES * 60, // seconds
    });
  }

  /** Verify + decode. Throws TokenExpiredError / InvalidTokenError to mirror PyJWT. */
  decodeAccessToken(token: string): AccessTokenClaims {
    try {
      return jwt.verify(token, this.config.JWT_SECRET, {
        algorithms: [this.config.JWT_ALGORITHM],
      }) as AccessTokenClaims;
    } catch (err) {
      if (err instanceof jwt.TokenExpiredError) {
        throw new TokenExpiredError((err as Error).message);
      }
      throw new InvalidTokenError((err as Error).message);
    }
  }

  /** Returns { raw, hash, expiresAt } — raw to the client, hash to the DB. */
  generateRefreshToken(): { raw: string; hash: string; expiresAt: Date } {
    const raw = crypto.randomBytes(48).toString('base64url');
    const hash = this.hashRefreshToken(raw);
    const expiresAt = new Date(
      Date.now() + this.config.REFRESH_TOKEN_LIFETIME_DAYS * 24 * 3600 * 1000,
    );
    return { raw, hash, expiresAt };
  }

  hashRefreshToken(raw: string): string {
    return crypto.createHash('sha256').update(raw).digest('hex');
  }
}
